### PR TITLE
[ENH] Extend forecasting benchmark to global setting

### LINKED ISSUE #6661: [ENH] Extend forecasting benchmark to global setting
As far as I know, currently the forecasting benchmarking (as well as evaluate) is not working for global setting. I.e., fitting a forecasting model on time series 1:n and evaluating it on n+1:m with n<m. 

@fkiraly please mention if it is already implemented.

### PR DESCRIPTION
**Pull request by XinyuWuu**
To close #6661 

Currently only a simple change is made to `evaluate`:
If `global_mode=True` passed to evaluate, `cv` (must be `InstanceSplitter`) will be used to split data in instance level and `cv_ht` (must be SingleWindowSplitter) will be used to split `y_test` from `cv` to `y_hist` and `y_true`, then `y_train, y_hist, y_true, X_train, X_test` will be used to evaluate global forecasters.
demo code snippet:
<details>

```py
from sktime.forecasting.naive import NaiveForecaster
from sktime.forecasting.pytorchforecasting import PytorchForecastingTFT
from sktime.utils._testing.hierarchical import _make_hierarchical
from sktime.benchmarking.forecasting import ForecastingBenchmark

data = _make_hierarchical(
    hierarchy_levels=(5, 100),
    max_timepoints=48,
    min_timepoints=48,
    n_columns=3,
)

x = data[["c0", "c1"]]
y = data["c2"].to_frame()


from sklearn.model_selection import KFold

from sktime.performance_metrics.forecasting import MeanSquaredPercentageError
from sktime.split import (
    CutoffFhSplitter,
    ExpandingWindowSplitter,
    InstanceSplitter,
    SingleWindowSplitter,
)
from sktime.split.base import BaseSplitter, BaseWindowSplitter

scorers = [MeanSquaredPercentageError()]

from sktime.forecasting.model_evaluation import evaluate

res = evaluate(
    PytorchForecastingTFT(**(PytorchForecastingTFT.get_test_params()[0])),
    cv=InstanceSplitter(KFold(2)),
    y=y,
    X=x,
    scoring=scorers,
    error_score="raise",
    global_mode=True,
    cv_ht=SingleWindowSplitter(fh=[1, 12], window_length=36),
)
print(res)
```

</details>
We might want to split data into multi subsets in instances level, and then split each subset into multi subsets again in time index level. 

We also need to extend `forecasting_validation` and `ForecastingBenchmark` to global setting.

@benHeid @fkiraly Any suggestions on the design?

### CODE REVIEW

**Comment by Xinyu-Wu-0000** [comment]
@benHeid The timeout error is caused by multiprocessing backend being not compatible with pytorch-forecasting models. I am sure why but it's blocking the tests so I skipped them. I also added a file lock as pytest fixture to enable running some tests sequentially to reduce memory usage.

**Comment by benHeid** [comment]
> @benHeid The timeout error is caused by multiprocessing backend being not compatible with pytorch-forecasting models. I am sure why but it's blocking the tests so I skipped them. I also added a file lock as pytest fixture to enable running some tests sequentially to reduce memory usage.

Mhm perhaps the models were to large and the runtime too long (10 epochs).. You could try to massively reduce the model size and the epochs. 

But. I think we should try to implement a different global forecaster that is not depending on pytorch-forecasting. So that this test is not depending on a soft-dependency. Perhaps, we need something like a global NaiveForecaster for such tests.

**Comment by Xinyu-Wu-0000** [comment]
> Mhm perhaps the models were to large and the runtime too long (10 epochs).. You could try to massively reduce the model size and the epochs.

I replaced TFT with DeepAR after #6551 was merged but it don't help. 
Normally it should be done in about 10s if we take a look at [test-full (3.10, ubuntu-latest)](https://github.com/sktime/sktime/actions/runs/10129926393/job/28010871806?pr=6774#logs). 
However with `multiprocessing` it blocks the `pytest sktime/forecasting/model_evaluation/tests/test_evaluate.py::test_evaluate_global_mode` for hours without any progress.



> But. I think we should try to implement a different global forecaster that is not depending on pytorch-forecasting. So that this test is not depending on a soft-dependency. Perhaps, we need something like a global NaiveForecaster for such tests.

If we simply refit `NaiveForecaster` on new data from `predict`, it would be global right?

**Comment by benHeid** [comment]
> > Mhm perhaps the models were to large and the runtime too long (10 epochs).. You could try to massively reduce the model size and the epochs.
> 
> 
> 
> I replaced TFT with DeepAR after #6551 was merged but it don't help. 
> 
> Normally it should be done in about 10s if we take a look at [test-full (3.10, ubuntu-latest)](https://github.com/sktime/sktime/actions/runs/10129926393/job/28010871806?pr=6774#logs). 
> 
> However with `multiprocessing` it blocks the `pytest sktime/forecasting/model_evaluation/tests/test_evaluate.py::test_evaluate_global_mode` for hours without any progress.
> 
> 
> 
> 
> 
> 
> 
> > But. I think we should try to implement a different global forecaster that is not depending on pytorch-forecasting. So that this test is not depending on a soft-dependency. Perhaps, we need something like a global NaiveForecaster for such tests.
> 
> 
> 
> If we simply refit `NaiveForecaster` on new data from `predict`, it would be global right? 

Or just use the y from predict in the prediction code.

**Comment by benHeid** [comment]
@XinyuWuu what is the current state of this PR? Are you still working on it? If not I would try to find some time working on it.

**Comment by benHeid** [comment]
@XinyuWuu, FYI. I finalised this PR. Now it should be ready for review.

**Comment by felipeangelimvieira** [comment]
Hi @XinyuWuu and @benHeid,

I have a quick question about the `cv_ht` argument name. I was wondering if it might be helpful to use a more descriptive name—perhaps something like `cv_global` or `y_test_splitter`—to make it clearer for users. Or even removing global and adopting this behaviour only if `cv_ht` is not None?

Also, thinking ahead, I noticed that the evaluation function has some redundancy in its arguments (e.g., cv_ht is only used when global=True; cv_X is not always used). Do you think we might encounter scenarios in the future where additional flags are needed? If so, it might be worth considering introducing a new entity, such as an "EvaluationStrategy." This could help simplify the design and make the evaluation process more intuitive for new users.

What do you think?

**Review by fkiraly — CHANGES_REQUESTED** [review]
This is a difficult PR to review - key API considerations.

My main comment is, I think it suffices to add a single argument, `cv_global`, which is an `InstanceSplitter`, and by default `None`.

* if `cv_global` is passed, the behaviour is the same as current `global_mode`, using `cv_global` to split instances.
* The parameter `cv_ht` is superfluous, since we are already passing `cv` which can be used in that case

This way we also avoid the type polymorphism on `cv_y` and `cv_X`.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

I think only minor changes necessary:

* in order for the docstrings to render corerctly, we need to use double backticks for code
* `cv_global` docstring fields need to say what the default is
* we should also clarify what exactly the training set for global foercasting is. This is not clear from the current docstring. For instance, only the test set of `cv_global` is being split
* I think we need to allow `sklearn` splitters as well, for `cv_global`. If they are passed, they get wrapped in `InstanceSplitter`. There is no need to require the user to do that, if they need an `InstanceSplitter` anyway (at least until we have other splitters with tag `split_type="instance"`. Concerely, I would say that the docstring should look like
`cv_global: sklearn splitter, or sktime splitter with tag split_type="instance"`

**Review by fkiraly — CHANGES_REQUESTED** [review]
@benHeid, the tests seems to be genuinely failing.

Further, should we not add the `cv_global` argument after `strategy` and `error_score`, to avoid breaking changes?

**Comment by benHeid** [comment]
> @benHeid, the tests seems to be genuinely failing.
> 
> 
> 
> Further, should we not add the `cv_global` argument after `strategy` and `error_score`, to avoid breaking changes?

True. Probably my merge of the main broke this. Let me have a look this evening.

**Comment by fkiraly** [comment]
no longer failing then - ready for review?

**Review by fkiraly — APPROVED** [review]

**Comment by benHeid** [comment]
@fkiraly are the errors in the cI caused by this PR?

**Comment by XinyuWuu** [comment]
I can't reproduce the error locally.

**Comment by XinyuWuu** [comment]
Seems to related to package versions. The _custom_showwarning method in sktime/utils/warnings.py causes infinite recursion when handling certain warnings. See #7903.

**Comment by benHeid** [comment]
Ok I am now convinced that this is not a bug of this pull request. Probably a race condition that occurs here: see #7906
